lib/Sema/CSDiag.cpp

```diff
@@ -2296,6 +2297,34 @@ bool FailureDiagnosis::diagnoseContextualConversionError(
     }
   }
 
+  // Diagnose passing '[T]' instead of a sequence of 'T' to a variadic argument
+  // of type 'T...'.
```

Comment:
It would be nice if we could implement this using the new diagnostics framework (see CSFix & CSDiagnostics), as we're currently trying to move stuff out of CSDiag.cpp.

What you'll want to do is add a new constraint fix to CSFix.h, and then add it as a viable fix within `repairFailures` when the locator's last element is an `ApplyArgToParam` and you're dealing with an array type being passed to a variadic param. Unfortunately I don't believe there's currently an easy way of accessing the parameter flags from `repairFailures`, the simplest solution would probably be to store them on the constraint locator – see https://github.com/hamishknight/swift/commit/62d97fc39dda86a1145fd4c93096be10fc30c9f7 for an example of that.

Then you'll want to add a case to `simplifyFixConstraint` where you try to match the array element types and record the fix if successful.

Reply:
@hamishknight sounds good, I'll look into doing that. Thanks for the example!